README.md
# MelloTrainer
A FiveM (GTA5 Modded Multiplayer) Trainer Made by <a href="https://github.com/TheStonedTurtle">TheStonedTurtle</a>. This is a server-side trainer written in Lua with a NUI written with HTML, CSS, and JS.<br/>
<br/>



<h1>Installation</h1>
<ol>
<li>Drag and drop the <i>mellotrainer</i> folder into the resources folder of your server.</li>
<li>Add <i>start mellotrainer</i> to your server.cfg file.</li>
<li>Restart your server.</li>
</ol>

<h2>Note</h2>
As of v0.4.0, players are required to be logged into the Steam client in order for the new vehicle/skin save system to work.

<h1>Controls</h1>
<table>
<tbody>
<tr><th>Key</th><th>Action</th></tr>
<tr><td>F1</td><td>Open/Close the trainer</td></tr>
<tr><td>Arrow Keys</td><td>Move up,down,left, and right respectively.</td></tr>
<tr><td>Enter</td><td>Select the current trainer option</td></tr>
<tr><td>Backspace</td><td>Go back to the previous menu</td></tr>
<tr><td>F2</td><td>Toggle No-Clip Mode</td></tr>
<tr><td>F3</td><td>Teleport to Current Way Point</td></tr>
<tr><td>Z</td><td>Toggle Between Big & Small Minimap</td></tr>
</tbody>
</table>



<h2>No Clip Controls</h2>
<table>
<tbody>
<tr><th>Key</th><th>Action</th></tr>
<tr><td>F2</td><td>Toggle No-Clip Mode</td></tr>
<tr><td>Shift</td><td>Switch No-Clip Movement Speed</td></tr>
<tr><td>Q</td><td>Move Upwards</td></tr>
<tr><td>Z</td><td>Move Downwards</td></tr>
<tr><td>W</td><td>Move Forwards</td></tr>
<tr><td>S</td><td>Move Backwards</td></tr>
<tr><td>A</td><td>Rotate Left</td></tr>
<tr><td>D</td><td>Rotate Right</td></tr>
</tbody>
</table>


<h1>Features</h1>
<ul>
    <li>Admin Menu w/ Working Permissions</li>
    <li>Admin Only Trainer Options (Time/Weather, syncs with online players)</li>
    <li>General Settings (Map Blips, Player Blips, Overhead Names, etc)</li>
    <li>Online Players Menu w/ Spectate & Teleport Option</li>
    <li>Player Death Messages</li>
    <li>Player Join/Leave Notifications</li>
    <li>Player Skin Changing and Customization</li>
    <li>Player Skin Save & Load System</li>
    <li>Player Toggle Options (God Mode, ETC)</li>
    <li>Vehicle Spawning & Spawning Options</li>
    <li>Vehicle Customization and Modifications</li>
    <li>Vehicle Save & Load System</li>
    <li>Weapon Spawning</li>
    <li>Weapon Attachments/Infinite ammo</li>
    <li>Voice Overlay & Options (Proximity/Channel/Toggle)</li>
    <li>Noclip Abilities with Admin Only Toggle</li>
</ul>


<h1>Credits:</h1>
<ul>
<li><a href="https://github.com/pongo1231/ScorpionTrainer">Scorpion Trainer</a> for a lot of the basic NUI functionality & structure.</li>
<li><a href="https://github.com/citizenfx/project-lambdamenu">Lambda Menu</a> which I used to convert useful functionality from C++ to Lua.</li>
<li><a href="https://forum.fivem.net/t/release-simple-speedometer/7846">Simple Speedometer</a> for the speedometer used by the trainer. (Black Outline)</li>
<li><a href="https://forum.fivem.net/t/release-scammers-script-collection-09-03-17/3313">Player Blips</a> for the basic blip system.</li>
</ul>


<h1>Development Information:</h1>
This section is only intended for people with LUA experience and a basic understanding of JSON/HTML Attributes. The below information should be used to help understand what every attribute does and how to add new dynamic menus and option using them.

<h2>Trainer Option Attributes:</h2>

<table>
<tbody>
<tr><th>Attribute</th><th>Explanation</th></tr>
<tr><td>data-action</td><td>The action to callback to lua via a NUICallback. Space delimited values with the first value being the name of the NUICallback.</td></tr>
<tr><td>data-hover</td><td>Exact same as data-action but this triggers when they change to the option instead of selecting the option.</td></tr>
<tr><td>data-state</td><td>Holds a "ON"/"OFF" value for toggle options.</td></tr>
<tr><td>data-toggle</td><td>Global boolean lua variable to sync data-state with.</td></tr>
<tr><td>data-sub</td><td>ID of the new menu to show when selected.</td></tr>
<tr><td>data-share</td><td>Information to share with the sub menu action options.(won't do anything unless data-sub is also specified)</td></tr>
<tr><td>data-shareid</td><td>Updates the submenu ID to this if it exists. Useful for ensuring that a menu that is used my multiple different options will return to the correct place within the trainer.</td></tr>
<tr><td>data-require</td><td>Used for permission checks for <b>data-action</b> and <b>data-sub</b> events. See Below</td></tr>
</tbody>
</table>


<h2>Trainer Div Attributes:</h2>
<table>
<tbody>
<tr><th>Attribute</th><th>Explanation</th></tr>
<tr><td>data-container</td><td>Prevents this div from being turned into a menu by JS.</td></tr>
<tr><td>data-parent</td><td>The ID of the parent element so if they try to go back</td></tr>
<tr><td>data-staticmenu</td><td>A menu that will be created from static JSON by JS. This will require updating JS so it is recommended you use data-dynamicmenu instead.</td></tr>
<tr><td>data-dynamicmenu</td><td>Holds the name of the NUI Callback that will return a JSON object to populate the current menu with (includes sub menus). See JSON format below.</td></tr>
<tr><td>data-sharedinfo</td><td>Usually added by JS from the data-share attribute of the option. This information will be appended to the end of every data-action and data-hover when requested.</td></tr>
</tbody>
</table>



<h2>JSON/Table Format</h2>

```json
{
	"menuName": "Example Text",
	"data": {
		"action": "String",
		"sub": "String",
		"state": "String"
	},
	"submenu": []
}
```


<table>
<tbody>
<tr><th>Atribute</th><th>Explanation</th><tr>
<tr>
	<td>menuName</td>
	<td>The text to show in the menu</td>
</tr>
<tr>
	<td>data</td>
	<td>An object containg key value pairs for all data-* attributes to be added to new menu option</td>
</tr>
<tr>
	<td>submenu</td>
	<td>An Array of of objects that are formatted in the exact same way as the current object. Used for creating linked sub menus.</td>
</tr>


<h2>JSON Editing</h2>
I created a website hosted on my github.io pages for editing the Mello Trainer JSON. It is not the best website and may have performance issues but it accomplishes the required task. Please read the above JSON information before trying/asking questions about this editor.

<a href="https://thestonedturtle.github.io/mellotrainer/mellotrainer.html" target="_blank">TheStonedTurtle.github.io</a></li>


<h2>Custom Privileges</h2>
To create custom privileges within mello trainer using data-require you can follow the below template. This works for data-action and data-sub but does not support data-hover events. <i>Note: This can be done without triggering a server event.</i>

<h3>Client.lua</h3>
Add the below to any client lua file.


```
-- Request Cop Status
RegisterNUICallback("requirecop", function(data, cb)
    TriggerServerEvent("mellotrainer:requestCopStatus")
end)

-- Recieve Cop Status
RegisterNetEvent("mellotrainer:copstatus")
AddEventHandler("mellotrainer:copstatus", function(status)
    if(status)then
        SendNUIMessage({customprivilegecheck = true})
    else
        drawNotification("~r~Permission Denied!")
    end     
end)
```

<h3>Server.lua</h3>

Add the below to any server lua file.


```
RegisterServerEvent('mellotrainer:requestCopStatus')
AddEventHandler('mellotrainer:requestCopStatus', function(id) 
    local result = false
    
    -- Logic to check if they are a cop here

    TriggerClientEvent("mellotrainer:copstatus",source,result)
end)
```

Changelog.md
<h1>v0.5.5 (09-24-2017)</h1>
Added new trainer features, new Setting Saving, and some general bug fixes/improvments



<h3>Changes</h3>
<ul>
    <li>Fixed Cosmetic Damage toggle showing scratches</li>
    <li>Updated <b>data-require</b> to work with <b>data-action</b>, was previously only working with <b>data-sub</b></li>
    <li>Fixed Skin Changing menu when loading a saved skin</li>
    <li>Fixed Voice Proximity Sync</li>
    <li>Fixed Voice Channel Menu and added Syncing</li>
    <li>Fixed incorrect submenu linking</li>
    <li>JSON Menu Name updates(credits @Matty45)</li>
</ul>

<h3>Additions</h3>
<ul>
	<li>Added Restore Apperance toggle (Default = ON)</li>
	<li>Added Local Setting Saving System (Trainer Toggles)</li>
	<li>Added Some MK2 Weapons and some of their attachments</li>
    <li>Added custom privledge information to the readme document</li>
</ul>



<h1>v0.5.0 (09-06-2017)</h1>
JSON Updates, JSON Editor, Setting Sync System, New Location Menu, Blip/Overhead System Rewrite, and many other code improvements.

<h3>Changes</h3>
<ul>
    <li>Majority of JSON moved to external files</li>
    <li>Blip & Overhead Name System Rewrite</li>
    <li>Potential fix for vehicles despawning</li>
    <li>Trainer should now show above other NUI elements</li>
    <li>PvP is now enabled by this trainer, will add config setting in future version</li>
    <li>Settings in cl_variables.lua now sync correctly to their default state</li>
    <li>Changed Map Blips to be on by default</li>
    <li>Updated requestInput() to return false if the default text is returned</li>
    <li>Updated vehicle settings being applied/retrieved when getting into a vehicle</li>
    <li>Modified the F3 Teleport to Marker functionality</li>
    <li>Many small bug fixes</li>
</ul>

<h3>Additions</h3>
<ul>
    <li>Added the entire Location Teleport menu, some places may not load correctly</li>
    <li>Added a Player Blip Names toggle, true by default</li>
    <li>Added Admin Kill and Teleport Player To Me Commands</li>
    <li>Custom JSON Loading (Editor: <a href="https://thestonedturtle.github.io/mellotrainer/mellotrainer.html" target="_blank">TheStonedTurtle.github.io</a>)</li>
    <li>Added all Post-Bikers Vehciles and Weapons to the menu (Thanks <a href="https://forum.fivem.net/u/Fenwick" target="_blank">Fenwick</a>)</li>
    <li>Z key will now toggle the minimap big/small. Can be removed by removing the if statement in `cl_settings.lua` under the `-- Toggle minimap on keypress` comment.</li>
</ul>



<h1>v0.4.5 (08-23-2017)</h1>
New Noclip Mode, Admin kick & temp ban, and some bug fixes. 

<h3>Changes</h3>
<ul>
    <li>Changed Control Keys to prevent controller conflictions</li>
    <li>Fixed SNS Pistol Menu issues</li>
    <li>Data saving can now be toggled in the sv_config.lua file</li>
    <li>Restructued all of the files for easier managment</li>
    <li>Fixed Time Forward/Backwards causing crashes</li>
    <li>Changed some menu colours</li>
</ul>

<h3>Additions</h3>
<ul>
    <li>Added a directory check to the data saving system</li>
	<li>Version check, the trainer will now let you know if you are running an outdated version. (Credits to EssentialMode)</li>
	<li>Added Noclip Mode, can be toggled with F2</li>
	<li>Added Admin Only Noclip toggle in sv_config.lua</li>
	<li>Added a Steam Only connection option which kicks any player not using steam. Toggleable in the sv_config.lua file</li>
	<li>Added Kick and Temp Ban to the admin menu</li>
</ul>




<h1>v0.4.0 (08-15-2017)</h1>
New vehicle and skin saving system, a few bug fixes.

<h3>Changes</h3>
<ul>
    <li>Fixed never-ending death messages when using RPDeath</li>
    <li>Fixed being able to open the trainer in the pause menu</li>
    <li>Fixed some vehicle modification names</li>
    <li>Potential fix for god mode not working</li>
    <li>Adjusted order of some NUI tables</li>
</ul>

<h3>Additions</h3>
<ul>
	<li>New data-saving.lua file for local storage, REQUIRES STEAM</li>
	<li>New Local Vehicle and Skin saving system</li>
	<li>New Menu refreshing system</li>
</ul>


<h1>V 0.3.0 (08-09-2017)</h1>

Many updates and code optimizations. New member <a href="https://github.com/WolfKnight98" target="_blank">WolfKnight98</a> added to the team.


<h3>Changes</h3>
<ul>
	<li>Fixed trainer controls</li>
	<li>Updated resource manifest version and native names</li>
	<li>More Player Blip and Overhead Name Optimizations</li>
	<li>Slightly updated Menu Spawn names</li>
	<li>Fixed No Ragdoll Mode</li>
	<li>Fixed Torque and Power Boost</li>
	<li>Fixed Player Join/Leave messages</li>
	<li>Player Death message optimizations</li>
	<li>Updated voice name overlay</li>
	<li>Localized config settings</li>
	<li>Code Restructure and Optimizations</li>
</ul>

<h3>Additions</h3>
<ul>
	<li>New config.lua file for settings</li>
	<li>New util.lua file for utility functions</li>
</ul>


<h1>V 0.2.3 (08-07-2017)</h1>

Minor bug fixes and player blip fixes.


<h3>Changes</h3>
<ul>
	<li>Fixed the player join and leave messages</li>
	<li>Fixed the player blip issues</li>
	<li>Minor CSS update for WolfKnight</li>
</ul>


<h1>V 0.2.2 (08-03-2017)</h1>

FXServer Compatibility & minor fixes.



<h3>Additions</h3>
<ul>
	<li>Added Sting Ray to the Animal Category</li>
</ul>

<h3>Changes</h3>
<ul>
	<li>Fixed the player joined message</li>
	<li>Fixed compatibility issues with FXServer</li>
</ul>





<h1>V 0.2.1 (07-08-2017)</h1>

Added variable in general.lua for admin-only trainer access.


<h3>Additions</h3>
<ul>
	<li>Added a Admin Only trainer toggle</li>
</ul>



<h1>V 0.2.0 (07-08-2017)</h1>

Bug fixes

<h3>Changes</h3>
<ul>
	<li>Fixed vehicle modification menu so it no longer applies a mod when entering the menu</li>
	<li>Fixed the vehicle power boost options</li>
	<li>Fixed the vehicle torque boost options</li>
</ul>



<h1>V 0.1.9 (07-07-2017)</h1>

Fixed an error with online players always being detected as offline.

<h3>Changes</h3>
<ul>

cl_admin.lua
--[[--------------------------------------------------------------------------
	*
	* Mello Trainer
	* (C) Michael Goodwin 2017
	* http://github.com/thestonedturtle/mellotrainer/releases
	*
	* This menu used the Scorpion Trainer as a framework to build off of.
	* https://github.com/pongo1231/ScorpionTrainer
	* (C) Emre Cürgül 2017
	* 
	* A lot of useful functionality has been converted from the lambda menu.
	* https://lambda.menu
	* (C) Oui 2017
	*
	* Additional Contributors:
	* WolfKnight (https://forum.fivem.net/u/WolfKnight)
	*
---------------------------------------------------------------------------]]


--[[------------------------------------------------------------------------
	Player Management 
------------------------------------------------------------------------]]--
function GetNetworkPlayers()
    local players = {}

    for i = 0, 31 do 
        if ( NetworkIsPlayerConnected( i ) ) then 
            table.insert( players, i )
        end 
    end 

    if ( next( players ) == nil ) then return nil else return players end 
end 

--[[ function GeneratePlayerBans( id )
	local bans = { 0.016667, 0.085, 0.17, 0.25, 0.5, 1, 2, 3, 4, 5, 8, 10, 15, 20, 30, 40, 60, 120, 240, 480 }
	local options = {}

	for k, v in pairs( bans ) do 
		local option = {
			[ "menuName" ] = tostring( v ) .. " hours",
			[ "data" ] = {
				[ "action" ] = "adminban " .. id .. " " .. v
			}
		}

		table.insert( options, option )
	end 

	return options 
end ]]--

function GeneratePlayerAdminMenus( id )
	local serverid = GetPlayerServerId( id )

	local kick = {
		[ "menuName" ] = "Kick",
		[ "data" ] = {
			[ "action" ] = "adminkick " .. serverid
		}
	}

	local kick_reason = {
		[ "menuName" ] = "Kick (specify reason)", 
		[ "data" ] = {
			[ "action" ] = "adminkick input " .. serverid
		}
	}

	local temp_ban = {
		[ "menuName" ] = "Temp Ban", 
		[ "data" ] = {
			[ "action" ] = "admintempban " .. serverid
		}
	}

	--[[ local banOptions = GeneratePlayerBans( serverid )
	local ban = {
		[ "menuName" ] = "Ban", 
		[ "data" ] = {
			[ "sub" ] = 3
		}, 
		[ "submenu" ] = banOptions
	} ]]--

	local kill = {
		[ "menuName" ] = "Kill player", 
		[ "data" ] = {
			[ "action" ] = "adminkill " .. serverid
		}
	}

	local teleport = {
		[ "menuName" ] = "Teleport to me",
		[ "data" ] = {
			[ "action" ] = "admintp " .. serverid
		}
	}

	local options = { kick, kick_reason, temp_ban, teleport, kill }

	return options 
end 

RegisterNUICallback( "playermanagement", function( data, cb ) 
    local players = GetNetworkPlayers()
    local validOptions = {}
 
    for k, v in pairs( players ) do
    	local playerOptions = GeneratePlayerAdminMenus( v )

        table.insert( validOptions, {
            [ "menuName" ] = GetPlayerName( v ),
            [ "data" ] = {
                [ "sub" ] = GetPlayerServerId( v )
            },
            [ "submenu" ] = playerOptions
        } )
    end
 
    local customJSON = "{}"

    if ( getTableLength( validOptions ) > 0 ) then
        customJSON = json.encode( validOptions, { indent = true } )
    end
 
    SendNUIMessage( {
        createonlineplayersmenu = true,
        menuName = "playermanagement",
        menudata = customJSON
    } )
   
    if ( cb ) then cb( "ok" ) end
end )

RegisterNUICallback( "adminkick", function( data, cb ) 
	if ( data.action == "input" ) then 
		local id = tonumber( data.data[3] )
		local reason = requestInput( "", 60 )

		if ( reason ) then 
			TriggerServerEvent( 'mellotrainer:adminKick', id, "Kicked: " .. reason )
		end 
	else 
		local id = tonumber( data.action )

		TriggerServerEvent( 'mellotrainer:adminKick', id, "Kicked: You have been kicked from the server." )
	end 
end )

RegisterNUICallback( "admintempban", function( data, cb ) 
	local id = tonumber( data.action )
	TriggerServerEvent( 'mellotrainer:adminTempBan', id )
end )

RegisterNUICallback( "adminkill", function( data, cb ) 
	local id = tonumber( data.action )
	TriggerServerEvent( 'mellotrainer:s_adminKill', id )
end )

RegisterNetEvent( 'mellotrainer:adminKill' )
AddEventHandler( 'mellotrainer:adminKill', function( data, cb ) 
	local ped = GetPlayerPed( -1 )

	if ( DoesEntityExist( ped ) and not IsEntityDead( ped ) ) then 
		SetEntityHealth( ped, 0 )
	end 
end )

RegisterNUICallback( "admintp", function( data, cb ) 
	local id = tonumber( data.action )
	TriggerServerEvent( 'mellotrainer:s_adminTp', id )
end )

RegisterNetEvent( 'mellotrainer:adminTp' )
AddEventHandler( 'mellotrainer:adminTp', function( adminSource ) 
	local me = GetPlayerPed( -1 )
	local target = GetPlayerFromServerId( adminSource )
	local targetPed = GetPlayerPed( target )

	local coords = GetEntityCoords( targetPed, true )
	local x, y, z = coords.x, coords.y, coords.z 
	z = z + 3.5

	RequestCollisionAtCoord( x, y, z )

	SetEntityCoordsNoOffset( me, x, y, z, false, false, true )
end )


--    _______ _                    ____        _   _                 
--   |__   __(_)                  / __ \      | | (_)                
--      | |   _ _ __ ___   ___   | |  | |_ __ | |_ _  ___  _ __  ___ 
--      | |  | | '_ ` _ \ / _ \  | |  | | '_ \| __| |/ _ \| '_ \/ __|
--      | |  | | | | | | |  __/  | |__| | |_) | |_| | (_) | | | \__ \
--      |_|  |_|_| |_| |_|\___|   \____/| .__/ \__|_|\___/|_| |_|___/
--                                      | |                          
--                                      |_|                          

-- Forward one hour
function getForwardTime()
	local hour = GetClockHours() + 1
	if(hour == 24)then
		hour = 0
	end
	return {h=hour,m=GetClockMinutes(),s=GetClockSeconds()}
end


-- Back one hour
function getReverseTime()
	local hour = GetClockHours() - 1
	if(hour < 0)then
		hour = 23
	end
	return {h=hour,m=GetClockMinutes(),s=GetClockSeconds()}
end



-- Update time
RegisterNetEvent('mellotrainer:updateTime')
AddEventHandler('mellotrainer:updateTime', function(h,m,s)
	--Citizen.Trace("Time Updated")
	NetworkOverrideClockTime(h,m,s)
end)




RegisterNUICallback("time", function(data, cb)
	local h,m,s
	local action = data.action
	if (action == "input") then
		local request = tonumber(requestInput("", 2))

		if (request == nil) or (request > 24) or (request < 0) then
			drawNotification("~r~You must specify a number from 0-24.")
			return
		else
			h=request
			m=0
			s=0
		end
	elseif(action == "forwards")then
		local time = getForwardTime()
		h=time.h
		m=time.m
		s=time.s
	elseif(action == "backwards")then
		local time = getReverseTime()
		h=time.h
		m=time.m
		s=time.s
	else
		h=tonumber(data.action) or 0
		m=0
		s=0
	end
	TriggerServerEvent("mellotrainer:adminTime", PlayerId(), h,m,s)
	drawNotification("~g~Time changed to "..tostring(h)..":"..string.format("%02d", tostring(m)))
end)



-- __          __        _   _                  ____        _   _                 
-- \ \        / /       | | | |                / __ \      | | (_)                
--  \ \  /\  / /__  __ _| |_| |__   ___ _ __  | |  | |_ __ | |_ _  ___  _ __  ___ 
--   \ \/  \/ / _ \/ _` | __| '_ \ / _ \ '__| | |  | | '_ \| __| |/ _ \| '_ \/ __|
--    \  /\  /  __/ (_| | |_| | | |  __/ |    | |__| | |_) | |_| | (_) | | | \__ \
--     \/  \/ \___|\__,_|\__|_| |_|\___|_|     \____/| .__/ \__|_|\___/|_| |_|___/
--                                                   | |                          
--                                                   |_|                          



-- Change Weather Type
function changeWeatherType(type, freezeToggle)
	SetOverrideWeather(type)
	if(freezeToggle)then
		ClearWeatherTypePersist()
		SetWeatherTypePersist(type)
	end
end


-- Blackout
RegisterNetEvent('mellotrainer:updateBlackout')
AddEventHandler('mellotrainer:updateBlackout', function(toggle)
	SetBlackout(toggle)
end)


-- Wind
RegisterNetEvent('mellotrainer:updateWind')
AddEventHandler('mellotrainer:updateWind', function(toggle, heading)
	if(toggle) then
		SetWind(1.0)
		SetWindSpeed(11.99);
		SetWindDirection(heading)
	else
		SetWind(0.0)
		SetWindSpeed(0.0);
	end
end)


-- Weather
RegisterNetEvent('mellotrainer:updateWeather')
AddEventHandler('mellotrainer:updateWeather', function(weatherString, toggle)
	SetOverrideWeather(weatherString)
	if(toggle)then
		ClearWeatherTypePersist()
		SetWeatherTypePersist(weatherString)
	end
end)


RegisterNUICallback("weatheroptions", function(data, cb)
	local action = data.action
	local newstate = data.newstate

	-- Notification Toggle Text
	local text = "~r~OFF"
	if(newstate) then
		text = "~g~ON"
	end


	if(action == "wind")then
		featureWeatherWind = newstate
		local heading = GetEntityHeading(PlayerPedId())
		TriggerServerEvent("mellotrainer:adminWind", PlayerId(), featureWeatherWind, heading)
		drawNotification("Wind "..text)
	elseif(action == "freeze")then
		featureWeatherFreeze = newstate
		drawNotification("Persistent Weather "..text.."~w~, Select weather to apply")
	elseif(action == "blackout")then
		featureBlackout = newstate
		TriggerServerEvent("mellotrainer:adminBlackout", PlayerId(), featureBlackout)
		drawNotification("Blackout "..text)
	end

end)

RegisterNUICallback("weather", function(data, cb)
	local weather = data.action

	TriggerServerEvent("mellotrainer:adminWeather", PlayerId(), weather, featureWeatherFreeze)

	drawNotification("Weather changed to ~g~" .. weather .. ".")
	cb("ok")
end)



RegisterNetEvent("mellotrainer:adminstatus")
AddEventHandler("mellotrainer:adminstatus", function(status)
	if(status)then
		SendNUIMessage({adminaccess = true})
	else
		drawNotification("~r~You are not an admin!")
	end		
end)

-- Only show menu if they are an admin
RegisterNUICallback("requireadmin", function(data, cb)
	TriggerServerEvent("mellotrainer:isAdmin")
end)

-- Wait until in game to trigger proper join
Citizen.CreateThread(function()
	while true do
		Citizen.Wait( 0 )

		if NetworkIsSessionStarted() then
			TriggerEvent( "mellotrainer:init" )
			return
		end
	end
end )